rvt-patch-llvm/src/main.rs

```diff
@@ -261,6 +283,70 @@ fn insert_call_at_head<'a>(
     builder.build_call(f, &args, "");
 }
 
+////////////////////////////////////////////////////////////////
+// Function transformation functions
+////////////////////////////////////////////////////////////////
+
+/// Apply Rust mangling rule to a function name like 'foo::bar'.
+/// Does not insert a unique hash at the end or the final 'E' character.
+fn rust_mangle(n: &str) -> String {
+    // todo: I believe that this can be implemented by an elegant chain of
+    // iterators - but I am not smart enough to figure out the ownership, etc.
+    // issues that rustc reports so I went with the following code instead.
+    let mut rs = vec!["_ZN".to_string()];
+    for x in n.split("::") {
+        rs.push(format!("{}{}", x.len(), x));
+    }
+    rs.concat()
+}
+
+/// Find a function whose name matches a Rust-mangled prefix
+fn find_mangled_function<'a>(module: &Module<'a>, prefix: &str) -> Option<FunctionValue<'a>> {
+    let prefix = rust_mangle(prefix);
+    let mut of = module.get_first_function();
+    while let Some(f) = of {
+        if f.get_name().to_bytes().starts_with(prefix.as_bytes()) {
+            return Some(f);
+        }
+        of = f.get_next_function()
+    }
+    return None;
+}
+
+/// Delete the body of a function
+fn delete_body<'ctx>(f: &FunctionValue<'ctx>) {
```

Comment:
This is similar to what `replace_def_with_dec` does. 
Note that you must also call `f.remove_personality_function()` to keep llvm happy.

Reply:
Good catch.
The goal of the new code is to replace the existing basic blocks with new code - so I don't want to remove personality functions, etc. in this case.

But reducing repeated code is good so I have refactored both functions as follows:

- Modified delete_body to use fun.get_basic_blocks based on your code in replace_def_with_dec
- Changed replace_def_with_dec to call delete_body.